FFmpegInterop/HEVCSampleProvider.cpp

```diff
@@ -0,0 +1,120 @@
+		NALUSampleProvider(stream, reader)
+	{
+		// Parse codec private data if present
+		if (m_stream->codecpar->extradata != nullptr && m_stream->codecpar->extradata_size > 0)
+		{
+			// Check the HEVC bitstream flavor
+			if (m_stream->codecpar->extradata_size > 3 && (m_stream->codecpar->extradata[0] || m_stream->codecpar->extradata[1] || m_stream->codecpar->extradata[2] > 1))
+			{
+				// hvcC config format
+				TraceLoggingWrite(g_FFmpegInteropProvider, "HEVCCodecPrivate", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+					TraceLoggingValue(m_stream->index, "StreamId"));
+
+				m_isBitstreamAnnexB = false;
+
+				HEVCConfigParser parser{ m_stream->codecpar->extradata, static_cast<uint32_t>(m_stream->codecpar->extradata_size) };
+				m_naluLengthSize = parser.GetNaluLengthSize();
+				tie(m_spsPpsData, m_spsPpsNaluLengths) = parser.GetSpsPpsData();
+			}
+			else
+			{
+				// Annex B format
+				TraceLoggingWrite(g_FFmpegInteropProvider, "AnnexBCodecPrivate", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+					TraceLoggingValue(m_stream->index, "StreamId"));
+
+				AnnexBParser parser{ m_stream->codecpar->extradata, static_cast<uint32_t>(m_stream->codecpar->extradata_size) };
+				tie(m_spsPpsData, m_spsPpsNaluLengths) = parser.GetSpsPpsData();
+			}
+		}
+	}
+
+	HEVCConfigParser::HEVCConfigParser(_In_reads_(dataSize) const uint8_t* data, _In_ uint32_t dataSize) :
+		m_data(data),
+		m_dataSize(dataSize)
+	{
+		// Validate parameters
+		WINRT_ASSERT(m_data != nullptr);
+		THROW_HR_IF(MF_E_INVALID_FILE_FORMAT, m_dataSize < MIN_SIZE);
+	}
+
+	uint8_t HEVCConfigParser::GetNaluLengthSize() const noexcept
+	{
+		return m_data[21] & 0x03 + 1;
+	}
+
+	tuple<vector<uint8_t>, vector<uint32_t>> HEVCConfigParser::GetSpsPpsData() const
+	{
+		vector<uint8_t> spsPpsData;
+		vector<uint32_t> spsPpsNaluLengths;
+
+		// Scan the parameter sets
+		uint32_t pos{ 22 };
+		const uint8_t parameterSetCount{ m_data[pos++] };
+		for (uint8_t i{ 0 }; i < parameterSetCount; i++)
+		{
+			// Get the NALU type and count
+			THROW_HR_IF(MF_E_INVALID_FORMAT, pos + sizeof(uint8_t) + sizeof(uint16_t) >= m_dataSize);
+			const uint8_t naluType{ static_cast<uint8_t>(m_data[pos++] & 0x3F) };
+			const uint16_t naluCount{ _byteswap_ushort(*reinterpret_cast<const uint16_t*>(m_data + pos)) };
+			pos += sizeof(uint16_t);
+
+			// Save the NALU data if it's VPS, SPS, or PPS
+			const bool copyNaluData{ naluType == NALU_TYPE_HEVC_VPS || naluType == NALU_TYPE_HEVC_SPS || naluType == NALU_TYPE_HEVC_PPS };
+			if (copyNaluData)
+			{
+				// Reserve estimated space now to minimize reallocations
+				spsPpsNaluLengths.reserve(spsPpsNaluLengths.size() + naluCount);
+			}
+
+			// Scan the NALUs
+			for (uint16_t j{ 0 }; j < naluCount; j++)
+			{
+				// Get the NALU length
+				uint32_t naluLength{ GetAVCNaluLength(m_data + pos, m_dataSize - pos, sizeof(uint16_t)) };
```

Comment:
>m_dataSize - pos [](start = 56, length = 16)

Need bounds check here and below (in "spsPpsData.insert(spsPpsData.end(), m_data + pos, m_data + pos + naluLength);").

Reply:
Handled by bounds check within GetAVCNaluLength